enhancements/microshift/microshift-telemetry.md

```diff
@@ -0,0 +1,338 @@
+OpenShift.
+
+Labels:
+* `_id`: cluster ID. UUID string.
+
+#### `cluster:memory_usage_bytes:sum`
+This metric reports the memory usage in percentage. Already supported in
+OpenShift.
+
+Labels:
+* `_id`: cluster ID. UUID string.
+
+#### `cluster:usage:resources:sum`
+This metric reports the number of Kubernetes objects by resource type. Already
+supported in OpenShift.
+
+Labels:
+* `_id`: cluster ID. UUID string.
+* `resource`: name of the kubernetes resource. Values can be `pods`,
+  `namespaces`, `services`, `ingresses.networking.k8s.io`,
+  `routes.route.openshift.io`,
+  `customresourcedefinitions.apiextensions.k8s.io`.
+
+#### `cluster:usage:containers:sum`
+This metric reports the number of running containers. Already supported in
+OpenShift.
+
+Labels:
+* `_id`: cluster ID. UUID string.
+
+#### `microshift_version`
+This metric reports information about MicroShift version. New metric not
+supported in OpenShift.
+
+Labels:
+* `_id`: cluster ID. UUID string.
+* `version`: MicroShift version. Possible values are `\d{1}.\d{1,2}.\d{1,2}`.
+* `deployment_type`: Describes deployment type. Possible values are `rpm`,
+  `ostree`, `bootc`.
+* `deployment_id`: In case `deployment_type` is `ostree`, this will hold the
+  ostree commit. In case of `bootc`, include image SHA. Format is a 64
+  character SHA.
+
+
+Using the names and labels listed above, each sample requiring a float64
+and an int64 value, all metrics combined are under 2KB.
+
+In order to keep metrics synchronized with those of OpenShift, in terms of API
+support, MicroShift will be using CI:
+* A job shall deploy MicroShift with Telemetry enabled.
+* A job will use the staging endpoint for Telemetry API. No added traffic
+  and no confusing data when querying.
+* At the end of the job a new check shall be performed: check there is data
+  belonging to that specific cluster in the staging environment.
+
+### Sending metrics
+OpenShift is sending metrics through this API every 4.5 minutes. This is done
+to overcome staleness in Prometheus server on the Telemetry backend. Stale data
+may yield gaps when retrieving it, making querying and dashboards more nuanced
+and harder to read.
+
+MicroShift, by its nature, needs to cover different use cases and potential
+limitations: resource constrained environments where computing power is scarce
+or limited, unreliable networks where outages are common, slow networks where
+bandwidth is not the best, private networks where all external traffic is
+audited and maybe limited, applications that are stable once they start,
+disconnected clusters, etc. The dynamic behavior that metrics help uncover is
+only one of the many use cases MicroShift has.
+For these reasons, and because the intention of this enhancement is to get the
+basic structure of a metrics reporting system, MicroShift will start using 1h
+intervals to send metrics, which is a good compromise between resource usage
+and data granularity.
```

Comment:
After reading it again, I'm not sure how sending metrics every hour helps with unreliable networking. It's quite the opposite in fact since there's less chance for the transfer to complete if it isn't attempted often. Expect if the system retries until it succeeds?

Reply:
By mentioning unreliable networks the enhancement refers to the possibility of having gaps in the metrics.
There is no mention of retries because we allow gaps to happen and we do not require continuous reporting, so if some request gets lost we can tolerate waiting until the next one.

Reply:
> There is no mention of retries because we allow gaps to happen and we do not require continuous reporting, so if some request gets lost we can tolerate waiting until the next one.

My point would be that if the system reports on the hour, fails because of connectivity issue and doesn't retry then you'll be missing data for 2h. And the same might happen on the next run.

Reply:
Yes, our initial idea was to have a one day report because Telemetry for MicroShift wont be widespread (the majority of the deployments are offline) and we are mainly interested in adoption and basic data. If we lose some of the write requests it is not critical, as there will be a new retry in the next hour. We should get 24 retries every day, that serves our current purpose for this version.
Gaps in any of these metrics are not something we need to avoid, even if they are long.